### ISSUE TITLE
Using logger in other files

### ISSUE THREAD

**Issue by jonydev** [minodes/sinatra-logger#7]
hi, @devyhia 
    my code is behind:
I create Sinatra logger in App which extend Sinatra::Base
![image](https://cloud.githubusercontent.com/assets/18723023/23987055/68ea3fea-0a62-11e7-8293-7d25c29fbb86.png)
    And in this SalesRefreshApp class or other middleware like OpportunityResource in the above picture, the logger can work
![image](https://cloud.githubusercontent.com/assets/18723023/23987106/bb847824-0a62-11e7-85ca-443e9c834193.png)
![image](https://cloud.githubusercontent.com/assets/18723023/23987129/fd417212-0a62-11e7-9c3f-ffbf8be16dd8.png)
    But in other class, like the OpportunityStore which new in OpportunityResource, the logger instance can't be found, even I required sinatra/logger
![image](https://cloud.githubusercontent.com/assets/18723023/23987160/229086d4-0a63-11e7-98a4-6b28bf014948.png)
    The error log is behind:
![image](https://cloud.githubusercontent.com/assets/18723023/23987184/4093e356-0a63-11e7-83b7-924b984b57b3.png)
    I'm thinking about that, in Rails, we can use Rails.log to use the log instance everywhere, but I don't know is there any method I can call like "Rails.log" to find the logger instance

**Comment by jonydev** [comment]
Hi, @devyhia 
    I've solve the problem, in your code of sinatra/logger, you set "SemanticLogger[self.class.name]"  which is the instance of logger to env["rack.logger"]. But the environment only can read in the class which extend Sinatra::Base, so we can't get the env in other class. 
    I doubt that why you set the SemanticLogger instance to env[“rack.logger”]? after you set it, is there other way to get the instance of logger except [“rack.logger”]?

**Comment by devyhia** [comment]
@jonydev `env["rack.logger"]` is available through the entire application and multiple requests/threads. Also, `env["rack.logger"]` and `env["rack.errors"]` call the log function, automatically (for all requests/errors). Refer to: http://stackoverflow.com/a/2240842/7647835
> All Rack apps have the call method that get's invoked with the HTTP Request env, if you check the call method of this middleware this is what happens:

def call(env)
  began_at = Time.now
  status, header, body = @app.call(env)
  header = Utils::HeaderHash.new(header)
  log(env, status, header, began_at)
  [status, header, body]
end

> But the environment only can read in the class which extend Sinatra::Base, so we can't get the env in other class.
^ You don't need the instance in the environment to access SinatraLogger.

I will post a running code shortly.

**Comment by devyhia** [comment]
@jonydev Could you test with the following:
```
require 'sinatra/logger'

class OppotunityStore
  ... includes go here ...
  include SemanticLogger::Loggable

  def initialize(http_client, settings)
    ... init goes here ...

    logger.info "OpportunityStore did initialize!"
  end
end
```

**Comment by jonydev** [comment]
oh, yeah, Thank you very much.
   I just read the source code of SemanticLogger, and get the point same as your answer.the 'env' object only given by param in call method in all Rack app, but can't be called in other class.
   But the SemanticLogger are the global instance that can be called in everywhere. 
   Thanks again, and I suggest that you can add some docs about how to use SemanticLogger.

**Comment by devyhia** [comment]
@jonydev Yes, will do. If you encounter any problem, please, ping us again! 👍

**Comment by jonydev** [comment]
@devyhia unexpected, SemanticLogger::Loggable does not work, but I just using SemanticLogger['SalesRefreshApp'].info, and it worked. Is there other thing I missed?

![image](https://cloud.githubusercontent.com/assets/18723023/23990352/bc6bdcd8-0a70-11e7-9045-29b59710e79a.png)

**Comment by jonydev** [comment]
@devyhia In the last issue, I saw the author add code in Rack App like this:
![image](https://cloud.githubusercontent.com/assets/18723023/23990702/d7d7eee8-0a71-11e7-900e-c13781cea7f4.png)
I think we should set the logger instance to @semantic_logger, you can see the code of SemanticLogger
![image](https://cloud.githubusercontent.com/assets/18723023/23990754/fde323b4-0a71-11e7-81d4-881cccee6729.png)

**Comment by devyhia** [comment]
Yes, but if you refer to: 
- http://www.rubydoc.info/gems/semantic_logger/2.9.2/SemanticLogger/Loggable
- https://github.com/rocketjob/semantic_logger/blob/master/lib/semantic_logger/loggable.rb

You would find that defining logger is not necessary. But, maybe, that is on static functions (i.e. `def self.myFunc`). I am re-checking this now.

**Comment by devyhia** [comment]
@jonydev The following code is working on my end. Both static and normal/object methods should work without setting a logger getter function. This is the whole purpose of `include ::SemanticLogger::Loggable`. Anywhere in your application, you could always use `SemanticLogger['Anywhere'].info`. However, `include ::SemanticLogger::Loggable` just injects a logger instance into your object/class with the class name.

```
require 'sinatra/base'
require 'sinatra/logger'

class ExternalClass
  include ::SemanticLogger::Loggable

  def foo
    logger.info "Foo"
  end

  def self.bar
    logger.info "Bar"
  end
end

class App < Sinatra::Base
  logger filename: "test.#{settings.environment}.log", level: :trace

  get '/' do
    logger.info "GET / REQUESTED :D"

    # Encapsulated Method
    obj = ::ExternalClass.new
    obj.foo

    # Static Method
    ::ExternalClass.bar
  end

  run! if app_file == $0
end
```

**Comment by jonydev** [comment]
@devyhia I used your code in one ruby file, and get this problem
![image](https://cloud.githubusercontent.com/assets/18723023/23993346/30ebb4ac-0a7b-11e7-8f23-df34db9c3ba4.png)

I'm thinking that you are write all code in Rack app, so the ::SemanticLogger::Loggable can find it self.class.name is "App", so you can run success, but my code are in different file, the class name is difference, so I can't get the instance

**Comment by jonydev** [comment]
@devyhia Hi, I made some test and I don't know the reason.
1.Add ExternalClass ruby file, and new ExternalClass in App class, the SemanticLogger::Loggable could be found, and the code run success.
2.Add ExternalClassSec and ruby file, new ExternalClassSec in App class, and class foosec method, foosec will call ExternalClass.foo method, the SemanticLogger::Loggable  can't be found in ExternalClass class, and return error.
![image](https://cloud.githubusercontent.com/assets/18723023/23994763/6765bf40-0a81-11e7-9077-3e404c5e063a.png)
![image](https://cloud.githubusercontent.com/assets/18723023/23994770/6c4d2516-0a81-11e7-928d-bae6577eeef7.png)
![image](https://cloud.githubusercontent.com/assets/18723023/23994853/bacf0b32-0a81-11e7-948a-b66c0f6351f2.png)
![image](https://cloud.githubusercontent.com/assets/18723023/23994866/c9d55406-0a81-11e7-9d40-0571b417b4e8.png)

**Comment by jonydev** [comment]
@devyhia I have found the reason. As I include ExternalClass, and call foo method in ExternalClassSec class,  in fact, It called ExternalClassSec.logger instead of ExternalClass.logger, as I haven't include SemanticLogger::Loggable in ExternalClassSec, so returns error, if I include SemanticLogger::Loggable, it will run success.
Another way, if I call ExternalClass.foo in ExternalClassSec class, it will call the class method of self.logger, then it will call ExternalClass.logger, as I included SemanticLogger::Loggable in ExternalClass, it run success.
I think this is a basic grammar of Ruby, and I didn't know it before.
